### PR TITLE
Add Int32TreeReader for smaller int values

### PR DESCRIPTION
**Pull request by tzembo**
~~Implements the `bson.Marshaler` or `bson.ValueMarshaler` interfaces for the ORC decimal, union, and map types so that they are marshaled into their correct BSON types. For example, decimal types are now represented as `{"$numberDecimal": XXX }` instead of being represented as the `{ int: {XXX}, scale: XXX }` structs that the `TreeReader` for these types return.~~

EDIT: Removed, this issue can be fixed by using custom BSON encoders when marshaling.

Also created a new `Int32TreeReader` that wraps the existing `IntegerTreeReader`. This new reader is for integers that can be represented as int32 values instead of int64 values (everything except long integers). This should allow short and int values to be marshaled into the `$numberInt` BSON value instead of `$numberLong`.

### CODE REVIEW

File: decimal.go [hunk 1]
```diff
@@ -4,6 +4,10 @@ import (
 	"io"
```
Comment (deafgoat) [inline decimal.go]: Nice work! Do we need to do anything special for marshaling `BinaryTreeReader` or `DateTreeReader`?

Comment (tzembo) [reply decimal.go]: Just checked both and they marshal correctly as they are now. `BinaryTreeReader` just returns a slice of bytes that then becomes a BSON `$binary` value, and `DateTimeReader` is an embedded struct that contains a `time.Time` value but marshals into the correct value:

`{"$date":{"$numberLong":"1577997435762"}}`

**Review by craiggwilson — COMMENTED** [review]
This looks nice. I have a feeling that getting the bson specific marshaling work accepted upstream will be a problem. The better way to handle this is to implement a `bsoncodec.Encoder` as a surrogate for these types. This means that the types don't need to change or have anything else added to them, we just delegate to the encoder by registering it into a registry.

We have done this already in a couple of places in ADL. You can see the internal/mongoserver/command package and the bson_decoders.go file there.

**Comment by deafgoat** [comment]
Yeah, we haven't really been trying to merge things upstream - this package itself is a fork of Alice's fork.

@tzembo can you pls file a ticket to create a PR for Alice's changes to the repo?

Didn't think of that, using a custom encoder is an excellent idea!

**Comment by tzembo** [comment]
@craiggwilson Looked into `bson.ValueEncoder` and seems exactly what I needed, working on that now.

File: decimal.go (hunk 1 above)
Comment (deafgoat) [reply decimal.go]: Nice.

**Review by deafgoat — APPROVED** [review]

**Comment by deafgoat** [comment]
Still LGTM.

**Review by craiggwilson — APPROVED** [review]

**Pull request merged**